\section{Callbacks}
Javascript ist eine ereignisgesteuerte Programmiersprache. Das bedeutet, anstatt auf blockierenden Code zu warten, führt Javascript Operationen weiter aus und reagiert wenn die Aktion fertig ist. Um sicherzustellen, dass Javascript abhängige Funktionen erst nach der Fertigstellung von asynchronen Operationen ausführt, wurde das Prinzip der \textbf{Callbacks} eingeführt. Doch wie regelt der Browser in welcher Reihenfolge der Code ausgeführt wird?

\subsection{Javascripts Laufzeitkonzept}
Dazu muss im groben das Verhalten des Javascript-Interpreters erklärt werden. Die wichtigsten Komponenten eines Javascript-Interpreters ist der Call stack und die Event Loop.\\

\noindent
Wenn ein Skript eine Funktion \textit{f} aufruft, fügt der Interpreter die Funktion zum Aufrufsstapel \textit{(Call stack)} hinzu. Wird nun eine Funktion \textit{g} innerhalb von \textit{f} aufgerufen, wird die innere Funktion weiter oben in den Call stack hinzugefügt. Wenn die Ausführung von \textit{f} fertig ist, wird diese Funktion aus dem Stapel entfernt und die Funktion, die sich weiter oben befindet, aufgerufen. Der Call stack verfolgt das Lifo-Prinzip (Last in - first out)\cite{javascript-interpreter}. Vor dem Ausführen des Beispiels muss folgend konfiguriert werden:

\begin{center}
    async-patterns$\,\to\,$ webpack.config.js
\end{center}

\begin{figure}[H]
\begin{lstlisting}[basicstyle=\small]
module.exports = {
    mode: 'development',
    entry: './src/modules/callbacks/introduction.ts',
    ...
}
\end{lstlisting}
\caption{Die webpack.config.js Datei befindet sich im Hauptverzeichnis des Projekts.}
\end{figure}

\begin{figure}[H]
\begin{lstlisting}[basicstyle=\small]
function h(z: string): void {
    console.log(z);
    console.log(new Error().stack); // (A)
}

function g(y: string): void {
    h(y + 'c'); // (B)
}

function f(x: string): void {
    g(x + 'b'); // (C)
}

f('a'); // (D)
\end{lstlisting}
\caption{Nutzung von abhängigen Funktionen in einem synchronen Programmiermodell.}
\end{figure}

\noindent
Anfänglich, wenn das Programm startet, ist der Call stack leer. Erst ab Zeile D mit dem Funktionsaufruf f('a') hat der Aufrufstapel einen Eingangspunkt. Der Einstiegspunkt befindet sich im globalen Geltungsbereich (\textit{global scope}):

\begin{itemize}
\item Globaler Geltungsbereich
\end{itemize}

\noindent
Nachdem die Funktion g(x + 'b') in Zeile C aufgerufen wurde, hat der stack zwei Eingangspunkte:

\begin{itemize}
\item Geltungsbereich innerhalb von Funktion \textit{f}
\item Globaler Geltungsbereich
\end{itemize}

\noindent
Nachdem Funktion h(y + 'c') aufgerufen wurden in Zeile B, hat der stack drei Eingangspunkte:

\begin{itemize}
\item Geltungsbereich innerhalb von Funktion \textit{g}
\item Geltungsbereich innerhalb von Funktion \textit{f}
\item Globaler Geltungsbereich
\end{itemize}

\noindent
Beim Ausführen des Codes zeigt die Konsole wie der Callstack vorgegangen ist:

\begin{figure}[H]
\begin{lstlisting}
abc
Error
    at h (stack_trace.js:2:17)
    at g (stack_trace.js:6:5)
    at f (stack_trace.js:9:5)
    at <global> (stack_trace.js:11:1)
\end{lstlisting}
\caption{Stellt die Anordnung der Aufrufe von unten nach oben dar.}
\label{stack-trace-error}
\end{figure}

\noindent
Javascripts Laufzeitkonzept lässt sich mit einem Warteschlangen-System vergleichen. Unabhängig davon wie zeitnah asynchrone Ereignisse wie timeouts oder Antworten aus einem Netzwerk ankommen, diese Prozesse werden erst in die Schlange gesetzt wenn der Stapel leer ist.

\begin{figure}[H]
\begin{lstlisting}[basicstyle=\small]
setTimeout(function() {
    console.log('second');
}, 10);

let num: number = 0;

while (num < 100000000) { // 100 Millionen Iterationen
    num = num + 1;
}

console.log('first');
// first
// second
\end{lstlisting}
\caption{Mit der setTimeout()-Funktion können Funktionsaufrufe verzögert werden.}
\label{First-timeout-example}
\end{figure}

\noindent
Eine Javascript Umgebung führt nur eine Aufgabe nacheinander aus. Während das Programm ausgeführt wird, läuft im Hintergrund eine Eventschleife die eventuell eintreffende Ereignisse registriert. Diese Schleife wird auch als \textbf{Event Loop} bezeichnet. Im Beispiel von Abbildung \ref{First-timeout-example} wird die setTimout() Funktion genutzt. SetTimeout() ist eine WEB API Methode, die der Browser zur Verfügung stellt. Hierbei wird bestimmt mit welcher Verzögerung in Millisekunden die innere Funktion in den Callstack hinzugefügt wird. Um genau zu sein, wird die innere Funktion nicht direkt in den Call stack hinzugefügt. Die Eventschleife registriert, dass ein Ereignis nach einer Zeitverzögerung ausgelöst wurde, und sorgt dafür dass diese Funktion in den Callstack hinzugefügt wird, sobald dieser Leer ist. Wann auch immer ein Javascript Programm läuft, prüft die Event Loop ob ein neues Ereignis ausgelöst wurde\cite{regular-event-loop}.\\

\noindent
Um eine Funktion nach einer asynchronen Operation auszuführen, können Funktionen geschachtelt werden. Dazu wird eine Funktionen als Argument angenommen und erst dann aufgerufen, wenn eine blockierende Aktion fertig ist. Doch wie genau funktioniert diese Schachtelung von Funktionen?

\subsection{Funktionsweise}

Sowohl in Javascript als auch in Typescript sind Funktionen als Erste-Klasse-Objekte zu sehen. Aus diesem Grund können sie an Variablen gebunden oder als Argumente in anderen Funktionen oder als Rückgabewert übergeben werden. Diese übergebenen Funktionen heißen Callbacks.\\

\noindent
Callbacks, auch Rückruffunktionen genannt, bildet das Kernkonzept der \textbf{funktionalen Programmierung} in Javascript\cite{callbacks-intro}. In der funktionalen Programmierung sind Funktionen als Werte zu betrachten. Der Code besteht aus kleineren Funktionen, die in höher geordneten Funktionen eingesetzt oder miteinander kombiniert werden können (Komposition). Dadurch ist der Code wiederverwendbar und weniger fehleranfällig. Javascript bietet mit Arrays die Möglichkeit Callbacks sinnvoll einzusetzen.

\begin{figure}[H]
\begin{lstlisting}[basicstyle=\small]
const numbers: number[] = [1, 2, 3, 4, 5, 6];

function isEven(x): boolean { 
  return x % 2 === 0; 
}

const evenNumbers = numbers.filter(isEven);
console.log(evenNumbers) // 2 4 6
\end{lstlisting}
\caption{Die Funktion filter() gibt ein neues Array zurück.}
\label{callbacks-with-arrays}
\end{figure}

\noindent
Die Methode filter() nimmt Elemente eines Arrays heraus, basierend einer Funktion die, die Filterkonditionen bestimmt. Filter() ruft die Callback Funktion einmalig für jedes Element auf. Dabei wird zuerst, dass Element aus der Liste entnommen und dann die Kondition geprüft. Wenn also eine Funktion als Parameter übergeben wird, wird diese nicht sofort ausgeführt. Eine Funktion die ein Callback als Argument annimmt, wird auch höherrangige Funktion \textit{(Higher Order Function)} genannt\cite{callbacks-example}.

\subsection{Beispiel}
In den jeweiligen Sektionen wird das gleiche Praxisbeispiel, in verschiedenen Ausprägungen angewendet und gegenübergestellt. Mit der REST-API \textbf{JSONPlaceholder} wird ein Endpunkt für die Datenabfrage definiert. Hierbei handelt es sich um eine Open Source Endstelle, die Beispieldaten für das Prototyping oder zum Testen zur Verfügung stellt. Je nach Anwendungsfall werden verschiedene Ausführungen von Anfragen an diese Endstelle gemacht, um die Code-Beispiele so praxisnah wie möglich zu halten.\\

\noindent
Das folgende Beispiel richtet sich nach dem funktionalen Programmierparadigma, da durch die Nutzungen von \textbf{puren Funktionen}, das Callback Prinzip besser zur Geltung kommt. Eine pure Funktion, ist eine Funktion, die beim selben Input jedes mal das gleiche Output wiedergibt. Zu erwähnen ist jedoch, dass das Prinzip der Callbacks auch im objektorientiertem Schema umgesetzt werden kann. 

\begin{figure}[H]
\begin{lstlisting}[basicstyle=\small]
module.exports = {
    mode: 'development',
    entry: './src/modules/callbacks/stories-usage.ts',
    ...
}
\end{lstlisting}
\caption{Für die folgenden Beispiele sollte ../callbacks/stories-usage.ts als Eingangsdatei in die webpack.config.js definiert werden.}
\end{figure}

\begin{figure}[H]
\begin{lstlisting}[basicstyle=\small]
function makeRequest(url, onSuccess, onFailure?): void {
    const req = new XMLHttpRequest();
    req.open('GET', url);

    req.onload = () => {
        if (req.status === 200) {
            fakeLatency(() => onSuccess(JSON.parse(req.response));
        }
    };

    req.onerror = () => onFailure(Error('Network Error'));

    req.send();
}

function fakeLatency(callback): void {
    setTimeout(callback, 3000 * Math.random());
}

function baseUrl(): string {
    return 'https://jsonplaceholder.typicode.com/posts/';
}

function getAllChapters(onSuccess, onFailure?): void {
    makeRequest(baseUrl(), onSuccess, onFailure);
}

function getChapter(chapter, onSuccess, onFailure?): void {
    makeRequest(baseUrl() + chapter.toString(), onSuccess, onFailure);
}
\end{lstlisting}
\end{figure}

\begin{figure}[H]\ContinuedFloat
\begin{lstlisting}[basicstyle=\small]
function createElm(innerHTML): HTMLElement {
    const div = document.createElement('div');
    div.innerHTML = innerHTML;
    document.body.appendChild(div);
    return div;
}

function spawn(content: any): void {
    if (content instanceof Array === false) {
        content = [content];
    }
    content.forEach(elm => {
        const snippet = document.createElement('div');
        snippet.innerHTML = '<h1>${elm.title}</h1>
                             <div>
                                <i>${elm.id}</i>
                             </div>
                             <p>${elm.body}.</p>';

        document.body.insertBefore(snippet, loadingIcon);
    });
}

function catchError(err) {
     createElm('Ooops! Error Occurred! ${err}');
}

function displayFinished(): void {
    loadingIcon.style.display = 'none';
    createElm('All done!');
}

const loadingIcon = createElm('<svg>..</svg>');
\end{lstlisting}
\caption{Durch das funktionale Programmierparadigma verhindert man die Abhängigkeit von externen Variablen und umgeht somit unerwünschte Seiteneffekte.}
\end{figure}

\noindent
Dabei sind die wichtigsten Funktionen makeRequest(), getAllChapters(), getChapter() und spawn(). MakeRequest() führt eine Http-Anfrage gegen einen definierten Endpunkt aus. Wenn der Status der Antwort 200 beträgt, wird im Success-Callback vorangeschreitet. Bei einem Fehler wird der Error-Callback aufgerufen. Sowohl getAllChapter() als auch getChapter() führen makeRequest() aus, mit der Möglichkeit im Erfolgs- und im Fehlerfall der Anfrage eine Aktion auszuführen. Spawn() nimmt ein Kapitel entgegen und lädt diesen in die DOM. Sollte ein Anwendungsfall sein, alle Kapitel der API anzufragen und bei Ankunft der Daten anzuzeigen, würde dies wie folgt aussehen:

\begin{figure}[H]
\begin{lstlisting}[basicstyle=\small]
getAllChapters(function(response) {
    spawn(response);
    displayFinished();
});
\end{lstlisting}
\caption{Zuerst wird der API-Aufruf ausgeführt und mit dem daraus resultierendem Objekt wird spawn() aufgerufen.}
\end{figure}

\noindent
In diesem Fall wird spawn() erst aufgerufen, wenn eine Antwort vom Endpunkt ankommt. Da Funktionen als Argument anderer Funktionen übergeben werden, ist es möglich diese \glqq{}später\grqq{} innerhalb des Funktionsrumpfes aufzurufen \textit{(\glqq{}Call-back\grqq{})}. Um multiple asynchrone Operationen nacheinander auszuführen, müssen Funktionen ineinandergeschachtelt werden. So wird die Ausführung sequentiell fortgesetzt. Als Beispiel wird Kapitel eins bis drei chronologisch aufgerufen:

\begin{figure}[H]
\begin{lstlisting}[basicstyle=\small]
console.log('before execution');
getChapter(1, function(response1) {
    spawn(response1);
    getChapter(2, function(response2) {
        spawn(response2);
        getChapter(3, function(response3) {
            spawn(response3);
            displayFinished();
        });
    });
});
console.log('after execution');
\end{lstlisting}
\caption{Aktionen wie die Übertragung von Daten in einem Netzwerk finden in einem eigenen Thread statt.}
\end{figure}

\noindent
Auffallend in diesem Beispiel ist der Grad der Einrückung, der mit jeder weiteren asynchronen Operationen steigt. Der Ausführungsstrang sieht hier wie folgt aus aus:

\begin{enumerate}
    \item Console.log('before execution') wird ausgeführt.
    \item getChapter(1) wird ausgeführt.
    \item Die anonyme Funktion innerhalb von getChapter(1) wird erst aufgerufen, wenn das Objekt aus der Netzwerkanfrage ankommt. Alle darunter abhängigen Funktionen werden dementsprechend noch nicht aufgerufen.
    \item Console.log('after execution') wird ausgeführt. Der Call stack ist nun leer.
    \item Die Antwort ist angekommen. Die anonyme Funktion ruft spawn() auf.
    \item getChapter(2) wird aufgerufen.
    \item ...
\end{enumerate}

\noindent
Alle weiteren asynchrone Operationen innerhalb von getChapter(1) sind voneinander abhängig und bleiben innerhalb der selben Zeitleiste. Asynchronität ist ansteckend. Eine Funktion, die zur asynchronen Verarbeitung ein Callback nutzt, operiert selbst asynchron. Wenn nun ein Großteil eines Programms aus ineinandergeschachtelten Funktionen besteht, kann diese Struktur zur erhöhten Fehleranfälligkeit führen. Dank der sechsten Ecmascript Version können Pfeil-Funktionen in Javascript genutzt werden. Diese sind syntaktisch kürzer als anonyme Funktionen. Wenn nun die fehlgeschlagenen Callbacks zusätzlich verarbeitet werden sollen, dann würde dies so aussehen:

\begin{figure}[H]
\begin{lstlisting}[basicstyle=\small]
getChapter(1, response1 => { // (*)
    spawn(response1);
    getChapter(2, response2 => { // (**)
        spawn(response2);
        getChapter(3, response3 => { // (***)
            spawn(response3);
            displayFinished();
        }, err3 => catchError(err3)); // (***)
    }, err2 => catchError(err2)); // (**)
}, err1 => catchError(err1)); // (*)
\end{lstlisting}
\caption{In der Praxis sollte bei einem Http-Aufruf der Fehlerfall immer abgedeckt sein.}
\label{Nested-Callback-with-catch-error}
\end{figure}

\noindent
Es bildet sich eine Verzweigung, die sich immer weiter nach rechts ausbreitet. Zudem ist es auf dem ersten Blick schwer zu erkennen welcher Error-Callback zu welchem Kapitelaufruf gehört.\\

\noindent
Callbacks ist nur eine konventionelle Bezeichnung, für die Nutzung von Javascripts Funktionen. Es gibt kein bestimmtes Schlüsselwort oder Objekt in der Javascript Sprache, dass Callback heißt. Anstatt sofort ein Wert zurückzugeben, bietet das Callback Prinzip die Möglichkeit Daten zu verarbeiten, die das Programm nicht sofort zur Verfügung hat. Es sollte jedoch nicht davon ausgegangen werden, dass alle höherrangigen Funktionen asynchron operieren (siehe Abb. \ref{callbacks-with-arrays}). Sie bilden lediglich Javascripts ersten Ansatz Daten asynchron zu verarbeiten. Bei mehreren abhängigen asynchronen Prozessen müssen Callbacks ineinandergeschachtelt werden. Mehrere Level von geschachtelten Callbacks führen zu einer schwer nachvollziehbaren Code-Struktur. Deshalb wird der Code in Abbildung \ref{Nested-Callback-with-catch-error} auch als \glqq{}Callback-Hell\grqq{} bezeichnet. Um aus der Callback-Hell zu entkommen, ebnete Ecmascript 6 den Weg der nativen Nutzung von \textbf{Promises}.



